src/coreclr/src/utilcode/pedecoder.cpp

```diff
@@ -1770,20 +1770,31 @@ void PEDecoder::LayoutILOnly(void *base, BOOL allowFullPE) const
                            PAGE_READONLY, &oldProtection))
         ThrowLastError();
 
-    // Finally, apply proper protection to copied sections
-    section = sectionStart;
-    while (section < sectionEnd)
+    // Finally, apply proper protection to copied sections    
+    for (section = sectionStart; section < sectionEnd; section++)
     {
         // Add appropriate page protection.
-        if ((section->Characteristics & VAL32(IMAGE_SCN_MEM_WRITE)) == 0)
+#if defined(CROSSGEN_COMPILE) || defined(TARGET_UNIX)
+        if (section->Characteristics & IMAGE_SCN_MEM_WRITE)
+            continue;
+
+        DWORD newProtection = PAGE_READONLY;
+#else
+        DWORD newProtection = section->Characteristics & IMAGE_SCN_MEM_EXECUTE ?
+            section->Characteristics & IMAGE_SCN_MEM_WRITE ?
+                PAGE_EXECUTE_READWRITE :
```

Comment:
We should not be leaving any pages mapped as RWX.

Reply:
I have seen this used in other cases.

https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/pal/src/map/virtual.cpp#L614

, so I assumed we mey need to keep fidelity with what PE says - if it is RWX section then it should be mapped to RWX memory.

In reality, I think I have never seen RWX actually happening here. I think R2R code is always RX, perhaps intentionally.  
I think for our purposes here we could just handle RWX case as RX here.

Reply:
The Windows PE loader ignores executable permissions for IL-only images. It will never map executable sections in IL images as executable, even if the PE file asks for it.

This was defense-in-depth security mitigation, implemented long time in .NET Framework days. It is probably not as important now compared to how important it used to be, but it would not hurt to keep it.

Reply:
The memory must be executable so that we could execute code. It does not need to be writeable though. Not for R2R stuff.
I have changed this to `PAGE_EXECUTE_READ` for executable sections regardless of writeability.